### ISSUE TITLE
Bug with deduplicating MISPObjects within an Event via breakOnDuplicate

### ISSUE THREAD

**Issue by github-germ** [MISP/MISP#6917]
2.4.137

I believe there is a bug with the new feature to dedup objects when ingested into an event as per #6838. This occurs when setting `breakOnDuplicate=True` while adding an object to a NEW event --  the event is created; however; the object is not added to the event.

I am attaching a test script. Usage is:
```
$ ./dedupObj.py -h
Usage: dedupObj.py -a/--authKey authkey [-b/--breakOnDuplicate] [-d/--duplicate] [-n/--newEvent]
  -a MISP user auth key
  -b sets object.breakOnDuplicate=True; otherwise False
  -d will attempt to ingest 2 of the same objects; otherwise only 1
  -n create a new Event; otherwise if Event.info exists, add
$ 
```

**Test Case Results**

There are more permutations that can be tested, but I think the key issue is just when setting `breakOnDuplicate` when creating a new event with objects, i.e. the objects are not included in the created event.

_These work well:_

1. Create new event with 1 object, `breakOnDuplicate=False`
   * `./dedupObj.py --authKey $authkey --newEvent`
   * Result: works well creating event with 1 object
2. Create new event with 2 identical objects, `breakOnDuplicate=False`
   * `./dedupObj.py --authKey $authkey --newEvent --duplicate`
   * Result: works well creating new event with 2 objects
3. Add to existing event with 1 object, `breakOnDuplicate=False`
   * `./dedupObj.py --authKey $authkey`
   * Result: works well adding 1 object to the existing event
4. Add to existing event with 1 object attempting to add a duplicate object, `breakOnDuplicate=True`
   * `./dedupObj.py --authKey $authkey --breakOnDuplicate`
   * Result: works well by not adding the duplicate object to the existing event

_These are not working as expected. Result exception is in ALL CAPS:_

1. Create new event with 1 object, `breakOnDuplicate=True`
   * `./dedupObj.py --authKey $authkey --newEvent --breakOnDuplicate`
   * Result: creates new event, NO OBJECT CREATED WITHIN THAT EVENT
2. Create new event with 2 identical objects, `breakOnDuplicate=True`
   * `./dedupObj.py --authKey $authkey --newEvent --duplicate --breakOnDuplicate`
   * Result: creates new event, NO OBJECT CREATED WITHIN THAT EVENT

[dedupObj.py.txt](https://github.com/MISP/MISP/files/5883063/dedupObj.py.txt)

**Comment by chrisinmtown** [comment]
what is `breakOnEvent`?

**Comment by github-germ** [comment]
> what is `breakOnEvent`?

Thanks. Just a typo in the description. Revising that now. The code uses the correct `breakOnDuplicate`

**Comment by github-germ** [comment]
@iglocska -- I suspect this may be an edge case with the `breakOnDuplicate` feature where `add_object`does not add objects when ingesting into a new event. Hoping you might find some time to take a look. I hear there might be a release coming soon, and it would be great to have this included. Many thanks!!

**Comment by github-germ** [comment]
@iglocska ... Hi Andras. I believe I've found two problems with the object dedup code. Hoping you can resolve for the next release.

1. As above, when creating a new event with objects, the event is created, but without any objects. The fix appears to be a `return` in the wrong place, best I can tell.
```
root@3sa-misp:/var/www/MISP/app/Model git --no-pager diff MispObject.php
diff --git a/app/Model/MispObject.php b/app/Model/MispObject.php
index 93ae308..69b7ad3 100644
--- a/app/Model/MispObject.php
+++ b/app/Model/MispObject.php
@@ -912,8 +912,9 @@ class MispObject extends AppModel
                         'title' => 'Object dropped due to it being a duplicate and breakOnDuplicate being requested for Event ' . $eventId,
                         'change' => 'Duplicate object found.',
                 ));
+                return true;
             }
-            return true;
+            // THIS BELONGS ABOVE: return true;
         }
         if (empty($log)) {
             $log = ClassRegistry::init('Log');
root@3sa-misp:/var/www/MISP/app/Model
```

2. The second issue is that when creating a new event with duplicated objects, they are not deduplicated, i.e. one can pass 2 identical objects in the new event, and both are ingested. I don't know the flow well enough yet. Originally, I was thinking this dedup could occur in `checkForDuplicateObjects` in `MispObjects.php`; however, after a bit more looking, I think it has to occur earlier in the flow. I defer to your expertise, and will learn more when I see your solution.

THANKS!

**Comment by iglocska** [comment]
Indeed, nice catch for issue 1!

I'll have a look at issue 2, but this week I'm mostly just giving trainings so it will take some time.

**Comment by github-germ** [comment]
@iglocska -- My guess for case 2 is that once the object is saved in `captureObject` and perhaps also in `saveObject` that `$this->__objectDuplicationCheckCache` needs to be updated to contain that newly saved object. I believe that updated cache will then catch any duplicate objects being passed as input into the new event in `checkForDuplicateObjects`.

HTH

Thanks.

**Comment by github-germ** [comment]
@iglocska -- Hi Andras, did you want anything else from me to get the fix for case 2?

**Comment by iglocska** [comment]
More hours in the day ;) Giving trainings all week, so a bit strapped for time.

**Comment by github-germ** [comment]
Hi @iglocska -- Thanks a ton!

My various test cases setting `breakOnDuplicate=True` on the object now all work well.

I am attaching my test script:  [dedupObj.py.txt](https://github.com/MISP/MISP/files/5997164/dedupObj.py.txt)

Usage for test case...
```
$ ./dedupObj.py --help
Usage: dedupObj.py -a/--authKey authkey [-b/--breakOnDuplicateObject] [-B/--BreakOnDuplicateEvent] [-d/--duplicate] [-n/--newEvent]
  -a MISP user auth key
  -b sets object.breakOnDuplicate=True; otherwise not set
  -B sets event.breakOnDuplicate=True (overrides -b); otherwise not set
  -d will attempt to ingest 2 of the same objects; otherwise only 1
  -n create a new Event; otherwise if Event.info exists, add
$ 
```


Am I understanding correctly that you also added the ability to set `breakOnDuplicate=True` at a higher level on the `MISPEvent` and that setting `object.breakOnDuplicate=True` would NOT be necessary?

If so, then dedup is working well when passing dups in a new event. However, I am seeing dedup not working when adding to an existing event.

* Case 1:
   * Dedup works well: Create a new event setting `event.breakOnDuplicate=True` (`object.breakOnDuplicate` is NOT set)
      ```
      ./dedupObj.py --authKey=KEY -duplicate -BreakOnDuplicateEvent --newEvent
      ```
* Case 2:
   * Does not dedup. Run this after Case 1 as it will add the dup objects into the existing event
      ```
      ./dedupObj.py --authKey=KEY --duplicate --BreakOnDuplicateEvent
      ```

   * Details of Case 2:
      1. Retrieve an existing event with `pymisp.search_index`
      2. Set `event.breakOnDuplicate=True` on that in-memory event
      3. Add an object with `event.add_object` to that in-memory event (without setting `object.breakOnDuplicate`) that already exists in the db, i.e. a dup.
      4.  Call `pymisp.update_event`
      5. **RESULT**: duplicate object exists in MISP


Please let me know what you think. Thanks again!

**Comment by iglocska** [comment]
Indeed, it should work on the event level now. I'll check, it might be only fixed for /events/add and not /events/edit

**Comment by github-germ** [comment]
Thanks! It does look like in `Event.php` https://github.com/MISP/MISP/commit/8bde7d01f488837f679b3ea7fd7c0eaa8fd6d716 revised `_add` but not `_edit`

**Comment by iglocska** [comment]
Yep indeed! Will try to add that asap, it's a bit silly this way.

**Comment by github-germ** [comment]
@iglocska -- Is there anything I can do from here to help? I think we're close.

**Comment by github-germ** [comment]
Hi @iglocska, Checking back to see if you have any time to fix [case 2 above](https://github.com/MISP/MISP/issues/6917#issuecomment-780688466)

**Comment by github-germ** [comment]
Hello @iglocska -- Any progress on the last bit on ensuring that `event.breakOnDuplicate=True ` also dedups correctly (as [above](https://github.com/MISP/MISP/issues/6917#issuecomment-781355313))?

**Comment by github-germ** [comment]
@iglocska -- not sure if you got to this with 2.4.143; please let me know. Thanks.
